Use paths mapping with tsconfig.json result to Cannot find module error

I'm using 3th party lib and have added  a path map in tsconfig,  but when i try to build it's give me 
Module not found: Error.

### Versions
```
Angular CLI: 1.7.0
Node: 9.4.0
OS: linux x64
Angular: 5.2.5
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

@angular/cli: 1.7.0
@angular/tsc-wrapped: 4.4.6
@angular-devkit/core: 0.4.0
@angular-devkit/schematics: 0.4.0
@ngtools/json-schema: 1.1.0
@schematics/package-update: 0.3.1
typescript: 2.6.2
webpack: 3.11.0

```
my tsconfing.json file:
```
{
  "compileOnSave": false,
  "compilerOptions": {
    "outDir": "./dist/out-tsc",
    "sourceMap": true,
    "declaration": false,
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "target": "es5",
    "typeRoots": [
      "node_modules/@types"
    ],
    "lib": [
      "es2017",
      "dom"
    ],
    "paths": {
      "ng-br-tools": [ "../dist/ng-br-tools" ]
    }
  }
}
```
my src/tsconfig.app.json file
```
{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "outDir": "../out-tsc/app",
    "baseUrl": "./",
    "module": "es2015",
    "types": []
  },
  "exclude": [
    "test.ts",
    "**/*.spec.ts"
  ]
}
```
### Repro steps
I put a 3 party lib in  dist/ng-br-tools  folder, and run `ng build`

### Observed behavior
```
 ng build
Date: 2018-02-16T11:52:30.826Z                                                          
Hash: caeb50b40575798f47f8
Time: 6547ms
chunk {inline} inline.bundle.js, inline.bundle.js.map (inline) 3.89 kB [entry] [rendered]
chunk {main} main.bundle.js, main.bundle.js.map (main) 22.3 kB [initial] [rendered]
chunk {polyfills} polyfills.bundle.js, polyfills.bundle.js.map (polyfills) 220 kB [initial] [rendered]
chunk {styles} styles.bundle.js, styles.bundle.js.map (styles) 14.6 kB [initial] [rendered]
chunk {vendor} vendor.bundle.js, vendor.bundle.js.map (vendor) 2.85 MB [initial] [rendered]

ERROR in ./src/app/app.module.ts
Module not found: Error: Can't resolve 'ng-br-tools' in '/home/103615563/projetos/ng-br-tools/src/app'
ERROR in ./src/app/app.component.ts
Module not found: Error: Can't resolve 'ng-br-tools' in '/home/103615563/projetos/ng-br-tools/src/app'
ERROR in ./src/app/services/sigep-web-cep.service.ts
Module not found: Error: Can't resolve 'ng-br-tools' in '/home/103615563/projetos/ng-br-tools/src/app/services'
ERROR in ./src/app/services/lst-estados-simples.service.ts
Module not found: Error: Can't resolve 'ng-br-tools' in '/home/103615563/projetos/ng-br-tools/src/app/services'
```

### Desired behavior
Paths should be resolved like expected in tsconfig


### Mention any other details that might be useful (optional)
This configurations works fine on  cli 1.6.8,  and 
if i create a syslink `ln -s dist/ng-br-tools node_modules/ng-br-tools
` to node_modules folder it's works too.

--- a/packages/@ngtools/webpack/src/paths-plugin.ts
+++ b/packages/@ngtools/webpack/src/paths-plugin.ts
@@ -20,7 +20,7 @@ export function resolveWithPaths(
   host: ts.CompilerHost,
   cache?: ts.ModuleResolutionCache,
 ) {
-  if (!request || !request.request) {
+  if (!request || !request.request || !compilerOptions.paths) {
     callback(null, request);
     return;
   }
@@ -32,21 +32,54 @@ export function resolveWithPaths(
   }
 
   // check if any path mapping rules are relevant
-  const isPathMapped = compilerOptions.paths && Object.keys(compilerOptions.paths)
-    .some(pattern => {
+  const pathMapOptions = [];
+  for (const pattern in compilerOptions.paths) {
       // can only contain zero or one
       const starIndex = pattern.indexOf('*');
       if (starIndex === -1) {
-        return pattern === request.request;
+        if (pattern === request.request) {
+          pathMapOptions.push({
+            partial: '',
+            potentials: compilerOptions.paths[pattern]
+          });
+        }
       } else if (starIndex === pattern.length - 1) {
-        return request.request.startsWith(pattern.slice(0, -1));
+        if (request.request.startsWith(pattern.slice(0, -1))) {
+          pathMapOptions.push({
+            partial: request.request.slice(pattern.length - 1),
+            potentials: compilerOptions.paths[pattern]
+          });
+        }
       } else {
         const [prefix, suffix] = pattern.split('*');
-        return request.request.startsWith(prefix) && request.request.endsWith(suffix);
+        if (request.request.startsWith(prefix) && request.request.endsWith(suffix)) {
+          pathMapOptions.push({
+            partial: request.request.slice(prefix.length).slice(0, -suffix.length),
+            potentials: compilerOptions.paths[pattern]
+          });
+        }
       }
-    });
+  }
 
-  if (!isPathMapped) {
+  if (pathMapOptions.length === 0) {
+    callback(null, request);
+    return;
+  }
+
+  if (pathMapOptions.length === 1 && pathMapOptions[0].potentials.length === 1) {
+    const onlyPotential = pathMapOptions[0].potentials[0];
+    let replacement;
+    const starIndex = onlyPotential.indexOf('*');
+    if (starIndex === -1) {
+      replacement = onlyPotential;
+    } else if (starIndex === onlyPotential.length - 1) {
+      replacement = onlyPotential.slice(0, -1) + pathMapOptions[0].partial;
+    } else {
+      const [prefix, suffix] = onlyPotential.split('*');
+      replacement = prefix + pathMapOptions[0].partial + suffix;
+    }
+
+    request.request = path.resolve(compilerOptions.baseUrl, replacement);
     callback(null, request);
     return;
   }

--- a/tests/e2e/tests/misc/module-resolution.ts
+++ b/tests/e2e/tests/misc/module-resolution.ts
@@ -1,10 +1,29 @@
-import { appendToFile, prependToFile } from '../../utils/fs';
+import { appendToFile, createDir, moveFile, prependToFile } from '../../utils/fs';
 import { ng, silentNpm } from '../../utils/process';
 import { updateJsonFile } from '../../utils/project';
 import { expectToFail } from '../../utils/utils';
 
 
 export default async function () {
+  await createDir('xyz');
+  await moveFile(
+    'node_modules/@angular/common',
+    'xyz/common'
+  );
+
+  await expectToFail(() => ng('build'));
+
+  await updateJsonFile('src/tsconfig.app.json', tsconfig => {
+    tsconfig.compilerOptions.paths = {
+      '@angular/common': [ '../xyz/common' ],
+    };
+  });
+  await ng('build');
+
+  await updateJsonFile('src/tsconfig.app.json', tsconfig => {
+    delete tsconfig.compilerOptions.paths;
+  });
+
   await prependToFile('src/app/app.module.ts', 'import * as firebase from \'firebase\';');
   await appendToFile('src/app/app.module.ts', 'firebase.initializeApp({});');
 
@@ -15,10 +34,6 @@ export default async function () {
   await silentNpm('install', 'firebase@4.9.0');
   await ng('build', '--aot');
   await ng('test', '--single-run');
-
-  // await prependToFile('src/app/app.module.ts', 'import * as firebase from \'firebase\';');
-  // await appendToFile('src/app/app.module.ts', 'firebase.initializeApp({});');
-  // await ng('build');
 
   await updateJsonFile('src/tsconfig.app.json', tsconfig => {
     tsconfig.compilerOptions.paths = {};

